File: leet-code/go/0021 merge-two-sorted-lists.go
package leet

// type ListNode struct {
// 	Val  int
// 	Next *ListNode
// }

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	if list1 == nil {
		return list2
	}

	s := list1
	b := list2

	if b != nil && s.Val > b.Val {
		tmp := *s
		s = b
		b = &tmp
	}

	for cur := s; b != nil && cur != nil; cur = cur.Next {
		if cur.Next == nil || b.Val < cur.Next.Val {
			tmp := *b
			b = cur.Next
			cur.Next = &tmp
		}
	}

	return s
}

File: leet-code/python/0021_merge_two_sorted_lists.py
# Definition for singly-linked list.
class ListNode:
  def __init__(self, val=0, next=None):
    self.val = val
    self.next = next
    
class Solution:
  def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
    stack = []
    while list1 is not None or list2 is not None:
      if list1 is not None and (list2 is None or list1.val < list2.val):
        stack.append(list1.val)
        list1 = list1.next
      else:
        stack.append(list2.val)
        list2 = list2.next
    
    if len(stack) == 0:
        return None

    head = ListNode(stack.pop(0))
    it = head
    while len(stack) > 0:
      it.next = ListNode(stack.pop(0))
      it = it.next
    
    return head